File: mlinit.lua
local assert = assert
local error = error
local getmetatable = getmetatable
local pairs = pairs
local pcall = pcall
local setmetatable = setmetatable
local math = math
local math_abs = math.abs
local math_type = math.type
local math_maxinteger = math.maxinteger
local math_mininteger = math.mininteger
local string = string
local string_format = string.format
local table = table
local table_pack = table.pack
local table_unpack = table.unpack

local function _Unit_EQUAL(t)
  return true
end
local function _Record_EQUAL(fields)
  return function(t)
    local a, b = t[1], t[2]
    for k,eq in pairs(fields) do
      if not eq({a[k], b[k]}) then
        return false
      end
    end
    return true
  end
end
local function _EQUAL(t)
  return t[1] == t[2]
end

local function _id(x)
  return x
end

local _exn_meta = {}
function _exn_meta:__tostring()
  return string_format("%s: %s", self.location or "<no location info>", self.tag[1])
end
local _Match_tag = { "Match" }
local _Match = setmetatable({ tag = _Match_tag }, _exn_meta)
local _Bind_tag = { "Bind" }
local _Bind = setmetatable({ tag = _Bind_tag }, _exn_meta)
local _Overflow_tag = { "Overflow" }
local _Overflow = setmetatable({ tag = _Overflow_tag }, _exn_meta)
local _Div_tag = { "Div" }
local _Div = setmetatable({ tag = _Div_tag }, _exn_meta)
local _Size_tag = { "Size" }
local _Size = setmetatable({ tag = _Size_tag }, _exn_meta)
local _Subscript_tag = { "Subscript" }
local _Subscript = setmetatable({ tag = _Subscript_tag }, _exn_meta)
local _Fail_tag = { "Fail" }
local function _Fail(message)
  return setmetatable({ tag = _Fail_tag, payload = message }, _exn_meta)
end

local function __exn_instanceof(e, tag)
  return getmetatable(e) == _exn_meta and e.tag == tag
end

local function _raise(x, location)
  local e = setmetatable({ tag = x.tag, payload = x.payload, location = location }, _exn_meta)
  error(e, 1)
end

-- Int
local function __Int_add(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  local z = x + y
  if y > 0 and z < x then
    _raise(_Overflow, "Int.+")
  elseif y < 0 and z < x then
    _raise(_Overflow, "Int.+")
  else
    return z
  end
end
local function __Int_sub(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  local z = x - y
  if y < 0 and z < x then
    _raise(_Overflow, "Int.-")
  elseif y > 0 and x < z then
    _raise(_Overflow, "Int.-")
  else
    return z
  end
end
local function __Int_mul(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  local z = x * y
  if (x ~= 0 and z // x ~= y) or (y ~= 0 and z // y ~= x) then
    _raise(_Overflow, "Int.*")
  else
    return z
  end
end
local function __Int_div(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Int.div")
  elseif x == math.mininteger and y == -1 then
    _raise(_Overflow, "Int.div")
  end
  return x // y
end
local function __Int_mod(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Int.mod")
  end
  return x % y
end
local function _Int_negate(x)
  assert(math_type(x) == "integer")
  if x == math_mininteger then
    _raise(_Overflow, "Int.~")
  end
  return - x
end
local function _Int_abs(x)
  assert(math_type(x) == "integer")
  if x == math_mininteger then
    _raise(_Overflow, "Int.abs")
  end
  return math.abs(x)
end

-- Word
local function __Word_div(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Word.div")
  elseif y > 0 then
    if x >= 0 then
      return x // y
    else -- x < 0
      local x1 = (x + math_maxinteger) + 1 -- x + 2^63
      local u1 = ((math_maxinteger % y) + 1) % y -- 2^63 % y
      local v1 = math_maxinteger // y
      if math_maxinteger % y == y - 1 then
        v1 = v1 + 1
      end
      -- v1 == 2^63 // y
      if x1 % y < y - u1 then
        return x1 // y + v1
      else
        return x1 // y + v1 + 1
      end
    end
  else -- y < 0
    if x >= 0 then
      -- x < 2^63 and y + 2^64 > 2^63
      -- Therefore, x // (y + 2^64) = 0
      return 0
    else -- x < 0
      -- (x + 2^64) // (y + 2^64)
      if x >= y then
        -- (x + 2^64) // (y + 2^64)
        return 1
      else
        return 0
      end
    end
  end
end
local function __Word_mod(x, y)
  assert(math_type(x) == "integer")
  assert(math_type(y) == "integer")
  if y == 0 then
    _raise(_Div, "Word.mod")
  elseif y > 0 then
    if x >= 0 then
      return x % y
    else -- x < 0
      local x1 = ((x + math_maxinteger) + 1) % y -- (x + 2^63) % y
      local u1 = ((math_maxinteger % y) + 1) % y -- 2^63 % y
      if y <= math_maxinteger // 2 + 1 then -- y <= 2^62
        -- x1 + u1 <= 2*(y-1) < 2^63
        return (x1 + u1) % y -- (x + 2^64) % y
      else
        -- 2^62 < y < 2^63
        -- math.maxinteger % y == math.maxinteger - y
        -- u1 = 2^63 - y <= 2^63 - (2^62 - 1) = 2^62 + 1
        -- 0 <= (2^63 - 1) - y <= (2^63 - 1) - (2^62 - 1) = 2^62
        -- 0 <= u1 - 1 <= 2^62
        -- 2^64 % y = ((2^63 % y) + (2^63 % y)) % y
        --          = ((2^63 - y) + (2^63 - y)) % y
        --          = (u1 + u1) % y -- u1
        local y2 = y // 2
        if x1 <= y2 then
          if u1 <= y2 then
            -- x1 + u1 <= 2 * (y // 2) <= y
            return (x1 + u1) % y
          else -- y2 < u1
            -- y - u1 <= y2
            -- -y2 <= x1 - (y - u1) <= y2
            return (x1 - (y - u1)) % y
          end
        else -- y2 < x1
          -- y - x1 <= y2
          if u1 <= y2 then
            return (u1 - (y - x1)) % y
          else -- y2 < u1
            local y3 = y - y2
            return ((x1 - y2) + (u1 - y3)) % y
          end
        end
      end
    end
  else -- y < 0
    if x >= 0 then
      -- x < 2^63 and y + 2^64 > 2^63
      return x
    else -- x < 0
      -- (x + 2^64) % (y + 2^64)
      if x >= y then
        return x - y
      else -- x < y
        return x
      end
    end
  end
end
local __Word_LT = math.ult

-- List
local _nil = { tag = "nil" }
local function _cons(t)
  return { tag = "::", payload = t }
end
local function _List_EQUAL(eq)
  local function go(a, b)
    local at, bt = a.tag, b.tag
    if at ~= bt then
      return false
    elseif at == "nil" then
      return true
    elseif eq({a.payload[1], b.payload[1]}) then
      return go(a.payload[2], b.payload[2])
    else
      return false
    end
  end
  return function(t)
    return go(t[1], t[2])
  end
end
local function _list(t)
  local xs = _nil
  for i = t.n, 1, -1 do
    xs = { tag = "::", payload = { t[i], xs } }
  end
  return xs
end

-- Ref
local function _ref(x)
  return { tag = "ref", payload = x }
end

-- Array
local function _Array_array(t)
  local n, init = t[1], t[2]
  if n < 0 then -- or maxLen < n
    _raise(_Size, "Array.array")
  end
  local t = { n = n }
  for i = 1, n do
    t[i] = init
  end
  return t
end
local function _VectorOrArray_fromList(xs)
  local t = {}
  local n = 0
  while xs.tag == "::" do
    n = n + 1
    t[n] = xs.payload[1]
    xs = xs.payload[2]
  end
  t.n = n
  return t
end
local function _VectorOrArray_tabulate(t)
  local n, f = t[1], t[2]
  if n < 0 then -- or maxLen < n
    _raise(_Size, "(Vector|Array).tabulate")
  end
  local t = { n = n }
  for i = 1, n do
    t[i] = f(i - 1)
  end
  return t
end
local function _VectorOrArray_sub(t)
  local a, i = t[1], t[2]
  if i < 0 or a.n <= i then
    _raise(_Subscript, "(Vector|Array).sub")
  end
  return a[i+1]
end
local function _Array_update(t)
  local a, i, x = t[1], t[2], t[3]
  if i < 0 or a.n <= i then
    _raise(_Subscript, "Array.update")
  end
  a[i+1] = x
  return nil
end

-- Vector
local function _Vector_EQUAL(eq)
  local function go(a, b)
    local n = a.n
    if n ~= b.n then
      return false
    end
    for i = 1, n do
      if not eq({a[i], b[i]}) then
        return false
      end
    end
    return true
  end
  return function(t)
    return go(t[1], t[2])
  end
end
local function _Vector_concat(xs)
  local n = 0
  local t = {}
  while xs.tag == "::" do
    local p = xs.payload
    local u = p[1]
    local m = u.n
    for i = 1,m do
      t[n + i] = u[i]
    end
    n = n + m
    xs = p[2]
  end
  t.n = n
  return t
end

-- Lua interface
local function _Lua_global(name)
  return _ENV[name]
end
local function _Lua_call(f)
  return function(v)
    return table_pack(f(table_unpack(v, 1, v.n)))
  end
end
local function _Lua_method(t)
  local self, name = t[1], t[2]
  return function(v)
    return table_pack(self[name](self, table_unpack(v, 1, v.n)))
  end
end
local function _Lua_newTable()
  return {}
end
local function _Lua_function(f)
  return function(...)
    local r = f(table_pack(...))
    return table_unpack(r, 1, r.n)
  end
end

File: mlinit.js
"use strict";
function _id(x) { return x; }
function _ref(x) { return { tag: "ref", payload: x }; }
const _nil = { tag: "nil" };
function _cons(a) { return { tag: "::", payload: a }; }
function _list(a) {
    var x = _nil;
    for (var i = a.length - 1; i >= 0; --i) {
        x = { tag: "::", payload: [a[i], x] };
    }
    return x;
}
function _Match_tag() {}
_Match_tag.prototype.name = "Match";
const _Match = new _Match_tag();
function _Bind_tag() {}
_Bind_tag.prototype.name = "Bind";
const _Bind = new _Bind_tag();
function _Div_tag() {}
_Div_tag.prototype.name = "Div";
const _Div = new _Div_tag();
function _Overflow_tag() {}
_Overflow_tag.prototype.name = "Overflow";
const _Overflow = new _Overflow_tag();
function _Size_tag() {}
_Size_tag.prototype.name = "Size";
const _Size = new _Size_tag();
function _Subscript_tag() {}
_Subscript_tag.prototype.name = "Subscript";
const _Subscript = new _Subscript_tag();
function _Fail_tag(payload) { this.payload = payload; }
_Fail_tag.prototype.name = "Fail";
function _Fail(payload) { return new _Fail_tag(payload); }
function _String_EQUAL(s, t) {
    if (s === t) { return true; }
    var n = s.length;
    if (n !== t.length) { return false; }
    for (var i = 0; i < n; ++i) {
        if (s[i] !== t[i]) {
            return false;
        }
    }
    return true;
}
const MIN_INT32 = -0x80000000;
const MAX_INT32 = 0x7fffffff;
function _Int_abs(x) {
    if (x < 0) {
        if (x === MIN_INT32) {
            throw _Overflow;
        }
        return -x;
    } else {
        return x;
    }
}
function _Int_negate(x) {
    if (x === MIN_INT32) {
        throw _Overflow;
    }
    return (-x)|0;
}
function __Int_add(x, y) {
    var z = x + y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z;
}
function __Int_sub(x, y) {
    var z = x - y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z;
}
function __Int_mul(x, y) {
    var z = x * y;
    if (z < MIN_INT32 || z > MAX_INT32) {
        throw _Overflow;
    }
    return z|0;
}
function __Int_div(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT32 && y === -1) {
        throw _Overflow;
    }
    return Math.floor(x / y)|0;
}
function __Int_mod(x, y) {
    if (y === 0) {
        throw _Div;
    }
    return x - Math.floor(x / y) * y;
}
function __Int_quot(x, y) {
    if (y === 0) {
        throw _Div;
    } else if (x === MIN_INT32 && y === -1) {
        throw _Overflow;
    }
    return (x / y)|0;
}
function __Int_rem(x, y) {
    if (y === 0) {
        throw _Div;
    }
    return (x % y)|0;
}
function __Word_div(x, y) {
    if (y === 0) {
        throw _Div;
    }
    return (x / y)>>>0;
}
function __Word_mod(x, y) {
    if (y === 0) {
        throw _Div;
    }
    return (x % y)>>>0;
}
const Math_abs = Math.abs;
const Math_imul = Math.imul;
function _call(f) {
    return function(args) {
        return f.apply(undefined, args);
    };
}
function _new(f) {
    return function(args) {
        return Reflect.construct(f, args);
    };
}
function _method(a) {
    var obj = a[0];
    var f = obj[a[1]];
    return function(args) {
        return f.apply(obj, args);
    };
}
function _encodeUtf8(s) {
    var encoder = new TextEncoder();
    return encoder.encode(s);
}
function _decodeUtf8(s) {
    var decoder = new TextDecoder();
    return decoder.decode(s);
}
function _exnName(e) { return _encodeUtf8(e.name); }
function _String_LT(a, b) {
    var i = 0;
    var m = a.length, n = b.length;
    while (i < m && i < n) {
        var x = a[i], y = b[i];
        if (x < y) {
            return true;
        } else if (x > y) {
            return false;
        }
        ++i;
    }
    return m < n;
}
function _String_append(a, b) {
    if (a.length === 0) { return b; }
    if (b.length === 0) { return a; }
    var c = new Uint8Array(a.length + b.length);
    c.set(a);
    c.set(b, a.length);
    return c;
}
function _String_concat(xs) {
    var n = 0;
    var xs0 = xs;
    while (xs0.tag === "::") {
        var s = xs0.payload[0];
        n += s.length;
        xs0 = xs0.payload[1];
    }
    var a = new Uint8Array(n);
    var m = 0;
    while (xs.tag === "::") {
        var s = xs.payload[0];
        a.set(s, m);
        m += s.length;
        xs = xs.payload[1];
    }
    return a;
}
function _String_concatWith(sep, xs) {
    var n = 0;
    var xs0 = xs;
    while (xs0.tag === "::") {
        var s = xs0.payload[0];
        n += s.length;
        xs0 = xs0.payload[1];
        if (xs0.tag === "::") {
            n += sep.length;
        }
    }
    var a = new Uint8Array(n);
    var m = 0;
    while (xs.tag === "::") {
        var s = xs.payload[0];
        a.set(s, m);
        m += s.length;
        xs = xs.payload[1];
        if (xs.tag === "::") {
            a.set(sep, m);
            m += sep.length;
        }
    }
    return a;
}
function _String_implode(xs) {
    var n = 0;
    var xs0 = xs;
    while (xs0.tag === "::") {
        ++n;
        xs0 = xs0.payload[1];
    }
    var a = new Uint8Array(n);
    var i = 0;
    while (xs.tag === "::") {
        a[i] = xs.payload[0];
        xs = xs.payload[1];
        ++i;
    }
    return a;
}
function _String_translate(f, s) {
    var m = s.length;
    var a = new Array(m);
    var n = 0;
    for (var i = 0; i < m; ++i) {
        var t = f(s[i]);
        a[i] = t;
        n += t.length;
    }
    var r = new Uint8Array(n);
    var l = 0;
    for (var i = 0; i < m; ++i) {
        r.set(a[i], l);
        l += a[i].length;
    }
    return r;
}
function _Array_array(t) {
    var n = t[0], init = t[1];
    if (n < 0) {
        throw _Size;
    }
    var a = new Array(n);
    a.fill(init);
    return a;
}
function _VectorOrArray_fromList(xs) {
    var a = [];
    while (xs.tag === "::") {
        a.push(xs.payload[0]);
        xs = xs.payload[1];
    }
    return a;
}
function _Vector_unsafeFromListRevN(n, xs) {
    var a = new Array(n);
    var i = n - 1;
    while (xs.tag === "::") {
        a[i] = xs.payload[0];
        xs = xs.payload[1];
        --i;
    }
    // i should be -1 here
    return a;
}
function _VectorOrArray_tabulate(t) {
    var n = t[0], f = t[1];
    if (n < 0) {
        throw _Size;
    }
    var a = new Array(n);
    for (var i = 0; i < n; ++i) {
        a[i] = f(i);
    }
    return a;
}
function _Vector_concat(xs) {
    var n = 0;
    var xs0 = xs;
    while (xs0.tag === "::") {
        var v = xs0.payload[0];
        n += v.length;
        xs0 = xs0.payload[1];
    }
    var a = new Array(n);
    var m = 0;
    while (xs.tag === "::") {
        var v = xs.payload[0];
        for (var i = 0; i < v.length; ++i) {
            a[m++] = v[i];
        }
        xs = xs.payload[1];
    }
    return a;
}
/*
type Trampoline<Result> = [true, Result] | exists X. [false, MLFunction<Result, X>, X]
interface MLFunction<Result, Arg> {
    (arg: Arg): Result;
    _MLTAIL_?: (arg: Arg) => Trampoline<Result>;
}
 */
function _wrap(f) {
    var F = function(x) {
        var result = f(x); // [true, Result] | exists X. [false, MLFunction<Result, X>, X]
        while (!result[0]) {
            var g = result[1];
            if ("_MLTAIL_" in g) {
                result = g._MLTAIL_(result[2]);
            } else {
                return g(result[2]);
            }
        }
        return result[1];
    }
    F._MLTAIL_ = f;
    return F;
}
function _function(f) {
    return function() {
        return f(arguments);
    };
}
